internal/featureflag/middleware.go

```
@@ -32,62 +35,47 @@ func Middleware(ffs Store, next http.Handler) http.Handler {
 // them.
 type flagSetFetcher struct {
 	ffs Store
-
-	once sync.Once
-	// Actor is the actor that was used to populate flagSet
-	actor *actor.Actor
-	// flagSet is the once-populated set of flags for the actor at the time of population
-	flagSet FlagSet
-}
-
-func (f *flagSetFetcher) fetch(ctx context.Context) FlagSet {
-	f.once.Do(func() {
-		f.actor = actor.FromContext(ctx)
-		f.flagSet = f.fetchForActor(ctx, f.actor)
-	})
-
-	currentActor := actor.FromContext(ctx)
-	if f.actor == currentActor {
-		// If the actor hasn't changed, return the cached flag set
-		return f.flagSet
-	}
-
-	// Otherwise, re-fetch the flag set
-	return f.fetchForActor(ctx, currentActor)
 }
 
-func (f *flagSetFetcher) fetchForActor(ctx context.Context, a *actor.Actor) FlagSet {
+func (f *flagSetFetcher) evaluateForActor(ctx context.Context, a *actor.Actor, flagName string) (flag *bool, err error) {
 	if a.IsAuthenticated() {
-		flags, err := f.ffs.GetUserFlags(ctx, a.UID)
-		if err == nil {
-			return FlagSet(flags)
+		if flag, err = f.ffs.GetUserFlag(ctx, a.UID, flagName); flag != nil {
+			setEvaluatedFlagToCache(flagName, a, *flag)
 		}
-		// Continue if err != nil
+		return flag, err
 	}
 
 	if a.AnonymousUID != "" {
-		flags, err := f.ffs.GetAnonymousUserFlags(ctx, a.AnonymousUID)
-		if err == nil {
-			return FlagSet(flags)
+		if flag, err = f.ffs.GetAnonymousUserFlag(ctx, a.AnonymousUID, flagName); flag != nil {
+			setEvaluatedFlagToCache(flagName, a, *flag)
 		}
-		// Continue if err != nil
+		return flag, err
 	}
 
-	flags, err := f.ffs.GetGlobalFeatureFlags(ctx)
-	if err == nil {
-		return FlagSet(flags)
+	if flag, err = f.ffs.GetGlobalFeatureFlag(ctx, flagName); flag != nil {
+		setEvaluatedFlagToCache(flagName, a, *flag)
 	}
+	return flag, err
+}
 
-	return FlagSet(make(map[string]bool))
+func EvaluateForActorFromContext(ctx context.Context, flagName string) (result bool) {
```

This function deserves a docstring. In particular you should call out that if the context doesn't have WithFlags set we silently return nil. In the old imlpementation we would panic (which isn't great, but atleast was noticeable)

---

Added docstring